runtime.c

```diff
@@ -519,8 +521,44 @@ Class class_setSuperclass(Class cls, Class newSuper)
 	CHECK_ARG(cls);
 	CHECK_ARG(newSuper);
 	if (Nil == cls) { return Nil; }
+
+	LOCK_RUNTIME_FOR_SCOPE();
+
+	safe_remove_from_subclass_list(cls);
+
 	Class oldSuper = cls->super_class;
 	cls->super_class = newSuper;
+
+	// The super class's subclass list is used in certain method resolution scenarios.
+	cls->sibling_class = cls->super_class->subclass_list;
+	cls->super_class->subclass_list = cls;
+
+	if (!class_isMetaClass(cls))
+	{
+		// Update the metaclass's superclass.
+		class_setSuperclass(cls->isa, newSuper->isa);
+	}
+	else
+	{
+		// newSuper is presumably a metaclass. Its isa will therefore be the appropriate root metaclass.
+		cls->isa = newSuper->isa;
+	}
+
+	// Make sure the superclass is initialized if we're initialized.
+	if (objc_test_class_flag(cls, objc_class_flag_initialized))
+	{
+		objc_send_initialize(newSuper);
```

Comment:
This is not safe.  `objc_send_initialize` expects to be called *without* the runtime lock being held and will both acquire and release it.  If it can't release the lock, then this can deadlock.